### ISSUE TITLE
GCR/Artifact registry authentication issue

### ISSUE THREAD

**Issue by s8525c** [renovatebot/renovate#15555]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.12.1

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

Gitlab 14.10.2

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Hello,

We host some of our docker images in (private) GCR and Artifact registry and renovate fails to look up dependencies for docker images.

Our config looks like this:

```yaml
DEBUG: Combined config
       "config": {
         "platform": "gitlab",
         "gitAuthor": "renovate-bot <renovate@acme.com>",
         "endpoint": "https://gitlab.acme.com/api/v4/",
         "token": "***********",
         "gitPrivateKey": "***********",
         "hostRules": [
           {"matchHost": "eu.gcr.io", "token": "***********"},
           {"matchHost": "europe-docker.pkg.dev", "token": "***********"},
           {"hostType": "github", "matchHost": "github.com", "token": "***********"}
         ],
         "repositories": [
           "acme/repo1",
           "acme/repo2",
           "acme/repo3"
         ],
         "requireConfig": true,
         "onboarding": true,
         "onboardingConfig": {
           "$schema": "https://docs.renovatebot.com/renovate-schema.json",
           "extends": ["local>acme/renovate-config"]
         }
       }
```

I also checked if the environment has access to these private registries and it does:

```shell
$ echo "Configure docker login..."
Configure docker login...
$ export GCR_TOKEN="$(gcloud auth print-access-token)"
$ echo "Troubleshoot gcr auth..."
Troubleshoot gcr auth...
$ gcloud container images list-tags europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image
DIGEST        TAGS                            TIMESTAMP
33b977c6f059  1.8.0,1.8.0.111032,1.8.latest   2022-05-11T13:59:42
58c2ed102c7d  1.8.0.108324                    2022-04-27T12:51:26
ad214e349973  1.8.0.104682                    2022-04-07T07:03:18
b155cb353851  1.8.0.103207                    2022-04-01T06:34:06
0796b49e58af  1.8.0.101572                    2022-03-23T10:21:46
...
$ gcloud container images list-tags europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image
DIGEST        TAGS                            TIMESTAMP
a42cb5c696ed  1.1.9,1.1.9.111048,1.1.latest   2022-05-11T14:40:13
d607f317fd34  1.1.9.110304                    2022-05-09T08:55:04
686787e599ba  1.1.9.109814                    2022-05-05T06:49:28
b87cf5c3b39f  1.1.9.109796                    2022-05-04T15:12:39
8c13e1cd07c2  1.1.9.108473                    2022-04-28T08:24:13
1d20b618fd51  1.1.8,1.1.8.108327              2022-04-27T12:55:16
8f0233876ee7  1.1.8.105425                    2022-04-08T14:31:56
1c379f83e613  1.1.7,1.1.7.104528              2022-04-06T12:42:08
910e5f9c78b6  1.1.7.104012                    2022-04-05T10:04:59
...
$ echo "Run renovate-bot..."
Run renovate-bot...
$ renovate --dry-run
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
...
```

I do not think the shared preset is relevant here, but I can post it if needed.

Thank you.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Found 2 package file(s) (repository=acme/my-awesome-repo)
 INFO: Dependency extraction complete (repository=acme/my-awesome-repo)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 8},
           "terraform": {"fileCount": 1, "depCount": 2}
         },
         "total": {"fileCount": 2, "depCount": 10}
       }
 WARN: Invalid registry response (repository=acme/my-awesome-repo)
       "apiCheckUrl": "https://europe-docker.pkg.dev/v2/gcp-project-id/acme-common/packer-image/tags/list?n=10000",
       "res": {
         "statusCode": 405,
         "body": "",
         "headers": {
           "content-length": "1458",
           "content-type": "text/html; charset=utf-8",
           "x-content-type-options": "nosniff",
           "x-frame-options": "SAMEORIGIN",
           "x-xss-protection": "0",
           "date": "Thu, 12 May 2022 07:00:55 GMT",
           "connection": "close"
         },
         "authorization": false
       }
DEBUG: Failed to get authHeaders for getTags lookup (repository=acme/my-awesome-repo)
 WARN: Invalid registry response (repository=acme/my-awesome-repo)
       "apiCheckUrl": "https://europe-docker.pkg.dev/v2/gcp-project-id/acme-common/terraform-image/tags/list?n=10000",
       "res": {
         "statusCode": 405,
         "body": "",
         "headers": {
           "content-length": "1458",
           "content-type": "text/html; charset=utf-8",
           "x-content-type-options": "nosniff",
           "x-frame-options": "SAMEORIGIN",
           "x-xss-protection": "0",
           "date": "Thu, 12 May 2022 07:00:55 GMT",
           "connection": "close"
         },
         "authorization": false
       }
DEBUG: Failed to get authHeaders for getTags lookup (repository=acme/my-awesome-repo)
DEBUG: Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image (repository=acme/my-awesome-repo, packageFile=.gitlab-ci.yml, dependency=europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image)
DEBUG: Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image (repository=acme/my-awesome-repo, packageFile=.gitlab-ci.yml, dependency=europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image)
DEBUG: Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image (repository=acme/my-awesome-repo, packageFile=.gitlab-ci.yml, dependency=europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image)
DEBUG: getLabels(https://index.docker.io, google/cloud-sdk, latest) (repository=acme/my-awesome-repo)
DEBUG: getManifestResponse(https://index.docker.io, google/cloud-sdk, latest) (repository=acme/my-awesome-repo)
DEBUG: Package releases lookups complete (repository=acme/my-awesome-repo)
       "baseBranch": "master"
DEBUG: branchifyUpgrades (repository=acme/my-awesome-repo)
DEBUG: 0 flattened updates found:  (repository=acme/my-awesome-repo)
DEBUG: Returning 0 branch(es) (repository=acme/my-awesome-repo)
DEBUG: config.repoIsOnboarded=true (repository=acme/my-awesome-repo)
DEBUG: packageFiles with updates (repository=acme/my-awesome-repo)
       "config": {
         "gitlabci": [
           {
             "packageFile": ".gitlab-ci.yml",
             "deps": [
               ...
               {
                 "depName": "europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image",
                 "currentValue": "1.8.0",
                 "replaceString": "europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image:1.8.0",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "image-name",
                 "depIndex": 1,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image",
                     "message": "Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/packer-image"
                   }
                 ],
                 "versioning": "docker"
               },
               {
                 "depName": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image",
                 "currentValue": "1.1.7",
                 "replaceString": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image:1.1.7",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "image-name",
                 "depIndex": 2,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image",
                     "message": "Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image"
                   }
                 ],
                 "versioning": "docker"
               },
               ...
               {
                 "depName": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image",
                 "currentValue": "1.1.7",
                 "replaceString": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image:1.1.7",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "image-name",
                 "depIndex": 5,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image",
                     "message": "Failed to look up dependency europe-docker.pkg.dev/gcp-project-id/acme-common/terraform-image"
                   }
                 ],
                 "versioning": "docker"
               },
               ...
             ]
           }
         ],
         ...
       }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
1. Is this the correct URL? `https://europe-docker.pkg.dev/v2/gcp-project-id/acme-common/packer-image/tags/list?n=10000`
2. Why is it returning a 405?

**Comment by rarkins** [comment]
BTW the gcloud log you send is helpful, but doesn't really prove a thing. If it's running `gcloud` commands instead of `docker` commands then it could be completely proprietary APIs it's using to get the list of tags. It doesn't prove whether their Docker API even supports tags listing (although I assume/hope it does).

**Comment by rarkins** [comment]
How is the `token` you configure for Renovate generated, and how long does it last for?

**Comment by s8525c** [comment]
Thanks for the very quick reply. I will try to answer your questions one by one:

- I had to redact the URL, but yes the URL is fine. We changed nothing in the config or in the project and it used to work. Also, that URL is something that renovate returns, it was not configured by us anywhere. I could dig into the logs and see when this problem occurred for the first time and post the relevant logs.
- Why 405? To be totally honest, I have no idea...
- We use `kaniko`, not `docker-in-docker`, but I think you are right about the `gcloud` approach. I will try to prove/disprove that `kaniko` can access those repositories and get back to you.
- The `token` is generated with `gcloud auth print-access-token` and is valid for 60 minutes. There are multiple ways to authenticate, and this is one way Google recommends (https://cloud.google.com/artifact-registry/docs/docker/authentication#token).
